src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -613,12 +613,52 @@ mongoc_topology_apply_scanned_srv_hosts (mongoc_uri_t *uri,
    return had_valid_hosts;
 }
 
+/*
+ *--------------------------------------------------------------------------
+ *
+ * mongoc_topology_should_rescan_srv --
+ * 
+ *      Checks whether it is valid to rescan SRV records on the topology.
+ *      Namely, that the topology type is Sharded or Unknown, and that
+ *      the topology URI was configured with SRV.
+ * 
+ *      If this returns false, caller can stop scanning SRV records
+ *      and does not need to try again in the future.
+ * 
+ *      NOTE: this method expects @topology's mutex to be locked on entry.
+ *
+ * --------------------------------------------------------------------------
+ */
+bool
+mongoc_topology_should_rescan_srv (mongoc_topology_t *topology) {
+   const char *service;
+
+   MONGOC_DEBUG_ASSERT (COMMON_PREFIX (mutex_is_locked) (&topology->mutex));
+
+   service = mongoc_uri_get_service (topology->uri);
+   if (!service) {
+      /* Only rescan if we have a mongodb+srv:// URI. */
+      return false;
+   }
+
+   if ((topology->description.type != MONGOC_TOPOLOGY_SHARDED) &&
+       (topology->description.type != MONGOC_TOPOLOGY_UNKNOWN)) {
+      /* Only perform rescan for sharded topology. */
+      return false;
+   }
```

Comment:
Why create a thread at all if the topology description isn't sharded or unknown? It seems like this check could be moved to `_mongoc_topology_background_monitoring_start` or `_start_scanner_if_needed`.